### PR TITLE
Check for context cancellation in more places within the winhttp transport layer before each I/O call.

### PR DESCRIPTION
**Pull request by ahsonkhan**
Fixes https://github.com/Azure/azure-sdk-for-cpp/issues/1107

~Todo: Figure out if its feasible to add unit tests, or if live tests are sufficient to cover the scenario.~ Existing tests cover it enough.

### CODE REVIEW

**Review by vhvb1989 — COMMENTED** [review]
Check for cancellation only before making I/O request. Just before writing or reading from network, or in between such operations.

Remove the checking from those places where we are just creating the request for winHttp or setting up the transport adapter

**Comment by ahsonkhan** [comment]
> Check for cancellation only before making I/O request. Just before writing or reading from network, or in between such operations.

I thought I did.

> Remove the checking from those places where we are just creating the request for winHttp or setting up the transport adapter

Can you give a specific example? I am guessing you mean the call to `WinHttpOpen`? Is that it?

File: sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp [hunk 1]
```diff
@@ -194,6 +194,8 @@ void GetErrorAndThrow(const std::string& exceptionMessage)
 
 void WinHttpTransport::CreateSessionHandle(std::unique_ptr<_detail::HandleManager>& handleManager)
 {
+  handleManager->m_context.ThrowIfCancelled();
```
Comment (vhvb1989) [inline sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp:197]: yes, This one might not be needed:

The WinHttpOpen function is the first of the WinHTTP functions called by an application. It initializes internal WinHTTP data structures and prepares for future calls from the application. When the application finishes using the WinHTTP functions, it must call WinHttpCloseHandle to free the session handle and any associated resources.

File: sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp [hunk 2]
```diff
@@ -246,6 +250,8 @@ void WinHttpTransport::CreateRequestHandle(std::unique_ptr<_detail::HandleManage
   const std::string& path = handleManager->m_request.GetUrl().GetRelativeUrl();
   HttpMethod requestMethod = handleManager->m_request.GetMethod();
 
+  handleManager->m_context.ThrowIfCancelled();
```
Comment (vhvb1989) [inline sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp:253]: This one I think we don't need it as well:

https://docs.microsoft.com/en-us/windows/win32/api/winhttp/nf-winhttp-winhttpopenrequest

The WinHttpOpenRequest function creates an HTTP request handle.

File: sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp [hunk 3]
```diff
@@ -407,6 +419,8 @@ int64_t WinHttpTransport::GetContentLength(
   // Get the content length as a number.
   if (requestMethod != HttpMethod::Head && responseStatusCode != HttpStatusCode::NoContent)
   {
+    handleManager->m_context.ThrowIfCancelled();
```
Comment (vhvb1989) [inline sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp:422]: I am not sure about this one.
I feel like queryHeaders will check the headers from memory from the response already read from network, so, it's not I/O.  I would remove this check

File: sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp [hunk 4]
```diff
@@ -430,6 +444,8 @@ std::unique_ptr<RawResponse> WinHttpTransport::SendRequestAndGetResponse(
     std::unique_ptr<_detail::HandleManager> handleManager,
     HttpMethod requestMethod)
 {
+  handleManager->m_context.ThrowIfCancelled();
```
Comment (vhvb1989) [inline sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp:447]: I wouldn't check for query

**Review by vhvb1989 — COMMENTED** [review]
Added the comments for when I don't think we need the Checking

**Review by RickWinter — COMMENTED** [review]
What is the perf impact of doing all these checks?

**Comment by ahsonkhan** [comment]
> What is the perf impact of doing all these checks?

It should be quite minimal since the bulk of the work here is I/O bound, and the cancellation check is expected to be really quick. We do this in curl as well. I can measure later to confirm, but that will require some setup.

File: sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp [hunk 5]
```diff
@@ -457,6 +465,8 @@ std::unique_ptr<RawResponse> WinHttpTransport::SendRequestAndGetResponse(
   // Allocate memory for the buffer.
   std::vector<WCHAR> outputBuffer(sizeOfHeaders / sizeof(WCHAR), 0);
 
+  handleManager->m_context.ThrowIfCancelled();
```
Comment (vhvb1989) [inline sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp:468]: No need to query

File: sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp [hunk 6]
```diff
@@ -476,6 +486,8 @@ std::unique_ptr<RawResponse> WinHttpTransport::SendRequestAndGetResponse(
   start = statusLineEnd + 1; // start of headers
   std::string responseHeaders = WideStringToString(std::wstring(start, last));
 
+  handleManager->m_context.ThrowIfCancelled();
```
Comment (vhvb1989) [inline sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp:489]: No need to query

File: sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp [hunk 7]
```diff
@@ -498,6 +510,8 @@ std::unique_ptr<RawResponse> WinHttpTransport::SendRequestAndGetResponse(
   uint16_t minorVersion = 0;
   ParseHttpVersion(httpVersion, &majorVersion, &minorVersion);
 
+  handleManager->m_context.ThrowIfCancelled();
```
Comment (vhvb1989) [inline sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp:513]: No need to query

File: sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp [hunk 8]
```diff
@@ -515,6 +529,8 @@ std::unique_ptr<RawResponse> WinHttpTransport::SendRequestAndGetResponse(
 
   HttpStatusCode httpStatusCode = static_cast<HttpStatusCode>(statusCode);
 
+  handleManager->m_context.ThrowIfCancelled();
```
Comment (vhvb1989) [inline sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp:532]: No need to query

File: sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp (hunk 5 above)
Comment (ahsonkhan) [reply sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp]: Good catches. Thanks.

**Review by RickWinter — APPROVED** [review]
This might be overly aggressive so we should evaluate perf.  Luckily its implementation detail and we have the liberty to add/remove in the future

**Review by vhvb1989 — APPROVED** [review]

**Pull request merged**